components/ee/ws-scheduler/pkg/scaler/driver.go

```diff
@@ -0,0 +1,380 @@
+				status.Count.Ghost++
+			}
+			log.WithField("delta", d).WithField("started", len(ids)).Info("controller requested ghost workspaces")
+		}
+	}
+}
+
+func (wspd *WorkspaceManagerPrescaleDriver) startGhostWorkspaces(ctx context.Context, count int, status workspaceStatus) (ids []string, err error) {
+	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
+	defer cancel()
+
+	ids = make([]string, count)
+	for i := 0; i < count; i++ {
+		if status.Count.Ghost+i > wspd.Config.MaxGhostWorkspaces {
+			log.WithField("limit", wspd.Config.MaxGhostWorkspaces).Warn("max number of ghost workspace reached")
+			return ids, nil
+		}
+
+		instanceUUID, err := uuid.NewRandom()
+		if err != nil {
+			return nil, err
+		}
+		instanceID := instanceUUID.String()
+		ids[i] = instanceID
+		metaUUID, err := uuid.NewRandom()
+		if err != nil {
+			return nil, err
+		}
+		metaID := metaUUID.String()
+
+		_, err = wspd.Client.StartWorkspace(ctx, &api.StartWorkspaceRequest{
+			Type: api.WorkspaceType_GHOST,
+			Id:   instanceID,
+			Metadata: &api.WorkspaceMetadata{
+				MetaId: metaID,
+				Owner:  wspd.Config.GhostOwner,
+			},
+			ServicePrefix: instanceID,
+			Spec: &api.StartWorkspaceSpec{
+				Admission:        api.AdmissionLevel_ADMIT_OWNER_ONLY,
+				Timeout:          "60m",
+				CheckoutLocation: "none",
+				FeatureFlags:     wspd.Config.FeatureFlags,
+				Git: &api.GitSpec{
+					Email:    "none@gitpod.io",
+					Username: "gitpod-ghost",
+				},
+				IdeImage: wspd.Config.IDEImage,
+				Initializer: &csapi.WorkspaceInitializer{
+					Spec: &csapi.WorkspaceInitializer_Empty{
+						Empty: &csapi.EmptyInitializer{},
+					},
+				},
+				WorkspaceLocation: "none",
+				WorkspaceImage:    wspd.Config.WorkspaceImage,
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ids, nil
+}
+
+func (wspd *WorkspaceManagerPrescaleDriver) stopGhostWorkspaces(ctx context.Context, ids []string) (err error) {
+	defer func() {
+		if err != nil {
+			err = xerrors.Errorf("cannot stop ghosts: %w", err)
+		}
+	}()
+
+	for _, id := range ids {
+		_, err := wspd.Client.StopWorkspace(ctx, &api.StopWorkspaceRequest{Id: id, Policy: api.StopWorkspacePolicy_NORMALLY})
+		if status.Code(err) == codes.NotFound {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (wspd *WorkspaceManagerPrescaleDriver) maintainWorkspaceStatus(ctx context.Context, counts chan<- workspaceStatus) error {
+	type workspaceState struct {
+		Started time.Time
+		Type    api.WorkspaceType
+	}
+
+	wss, err := wspd.Client.GetWorkspaces(ctx, &api.GetWorkspacesRequest{})
+	if err != nil {
+		return err
+	}
+
+	state := make(map[string]workspaceState)
+	produceStatus := func() workspaceStatus {
+		var res workspaceStatus
+		res.DeletionCandidates = make([]string, 0, len(state))
+		for id, s := range state {
+			switch s.Type {
+			case api.WorkspaceType_GHOST:
+				res.Count.Ghost++
+			case api.WorkspaceType_PREBUILD:
+				res.Count.Prebuild++
```

Comment:
We have configurations in which prebuilds are running in a separate NodePool, and deleting them would degrade the service without any help in scheduling workspaces.

Worth considering here?

Reply:
at the moment none of the controllers works specifically on prebuilds. This is more of a "we might need this in the future if we want to speed up prebuilds in the same way" kind of thing.

We could even remove this for now to avoid confusion.

Reply:
:+1:
Also, I don't get why `REGULAR` workspaces are added to `DeletionCandidates` as well. `DeletionCandidates` is passed to `stopGhostWorkspaces`, which in turn just stops them without any filtering for `type == GHOST` like the name suggests.

Reply:
that is a 🐛 

good catch!